File: app/javascript/applications/CharKeeperApp/helpers/pointBuy.js
// PH 2024 p.38 point buy, the only ability-score method TLC creation offers.
// Mirrors app/platforms/tlc/point_buy.rb, which is what actually gates a create:
// this side only keeps the form honest. Costs are cumulative from 8 and the
// 13->14 / 14->15 steps cost 2, so callers must read the table instead of
// counting one point per step.
export const POINT_BUY_COST = { 8: 0, 9: 1, 10: 2, 11: 3, 12: 4, 13: 5, 14: 7, 15: 9 };
export const POINT_BUY_BUDGET = 27;
export const POINT_BUY_MIN = 8;
export const POINT_BUY_MAX = 15;

// Every ability at the floor: the spread a fresh form starts from.
export const pointBuyFloor = () => ({
  str: POINT_BUY_MIN, dex: POINT_BUY_MIN, con: POINT_BUY_MIN,
  int: POINT_BUY_MIN, wis: POINT_BUY_MIN, cha: POINT_BUY_MIN
});

// NaN, not a silent 0, for a score off the table: every comparison against NaN
// is false, so an unpriceable spread can never read as affordable.
export const pointBuySpent = (abilities) =>
  Object.values(abilities).reduce((total, score) => total + POINT_BUY_COST[score], 0);

export const pointBuyRemaining = (abilities) => POINT_BUY_BUDGET - pointBuySpent(abilities);

// Can `slug` move by `step` (+1/-1) without leaving 8..15 or overdrawing the budget?
export const canPointBuyChange = (abilities, slug, step) => {
  const score = abilities[slug] + step;

  if (score < POINT_BUY_MIN || score > POINT_BUY_MAX) return false;

  return pointBuyRemaining({ ...abilities, [slug]: score }) >= 0;
};

File: app/platforms/tlc/point_buy.rb
# frozen_string_literal: true

module Tlc
  # PH 2024 p.38 point buy — the only ability-score method TLC character creation
  # offers (plan Phase A2: "level 3 default, point-buy only"). Costs are
  # cumulative from 8, and the 13->14 and 14->15 steps cost 2 rather than 1, so
  # callers must read the table instead of counting one point per step.
  #
  # Scores here are pre-species/background: the creation contract validates the
  # spread the player bought, before any boost is layered on.
  module PointBuy
    COST = { 8 => 0, 9 => 1, 10 => 2, 11 => 3, 12 => 4, 13 => 5, 14 => 7, 15 => 9 }.freeze
    BUDGET = 27
    MIN = COST.keys.min
    MAX = COST.keys.max

    # nil, never 0, when a score is off the table: an unpriceable spread has to
    # read as unaffordable rather than free.
    def self.spent(abilities)
      costs = abilities.values.map { |score| COST[score.to_i] }
      costs.include?(nil) ? nil : costs.sum
    end

    def self.affordable?(abilities)
      total = spent(abilities)

      !total.nil? && total <= BUDGET
    end
  end
end
